### PR TITLE
Improve error handling

### LINKED ISSUE #24: Reject with error
It is good practice to reject promises with an error object because they have stack traces.

[When rejecting because of a bad status code](https://github.com/mzabriskie/axios/blob/d93df704a269d3cbf8f00bcd4b2b2489d5b74f49/lib/adapters/http.js#L73), I suggest rejecting with an error object, e.g. `BadStatusCodeError`. You could then attach the response object to the error object so it can be accessed in a rejection handler.

### PR DESCRIPTION
**Pull request by nickuraltsev**
This PR fixes #24 and #242. 

**Breaking change:** Promise is now rejected with an `Error` rather than the response object.

### CODE REVIEW

File: lib/core/settle.js [hunk 1]
```diff
@@ -13,6 +15,8 @@ module.exports = function settle(resolve, reject, response) {
   if (!response.status || !validateStatus || validateStatus(response.status)) {
     resolve(response);
   } else {
-    reject(response);
+    var error = createError('Request failed with status code ' + response.status, response.config);
+    error.response = response;
```
Comment (mzabriskie) [inline lib/core/settle.js]: I would change this to pass `response` in as 4th arg to `createError`.

**Comment by romainneutron** [comment]
This would have required a note in the `UPGRADE_GUIDE.md` file

**Comment by misund** [comment]
Breaking changes should mean a major version change.

**Comment by mzabriskie** [comment]
@romainneutron https://github.com/mzabriskie/axios/blob/master/UPGRADE_GUIDE.md#error-handling

**Comment by mzabriskie** [comment]
@misund once we hit 1.0 this will definitely be the case. Until then as the API becomes more hardened, breaking changes will be released as new minor versions. There is a note in the README regarding this https://github.com/mzabriskie/axios#semver.

**Comment by romainneutron** [comment]
Thanks

**Comment by woodb** [comment]
Could you please note that this is a breaking change in the CHANGELOG?

**Pull request merged**